### PR TITLE
Implement otel spanlinks api

### PR DESCRIPTION
**Pull request by jackshirazi**
## What does this PR do?
Implements the span links part of the OTel bridge

## Checklist
- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] I have made corresponding changes to the documentation
- [ ] This is a bugfix
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] I have added tests that would fail without this fix
- [ ] This is a new plugin
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] My code follows the [style guidelines of this project](https://github.com/elastic/apm-agent-java/blob/main/CONTRIBUTING.md#java-language-formatting-guidelines)
  - [ ] I have made corresponding changes to the documentation
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/main/CONTRIBUTING.md#testing) pass locally with my changes
  - [ ] I have updated [supported-technologies.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/docs/supported-technologies.asciidoc)
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] Added an instrumentation plugin? Describe how you made sure that old, non-supported versions are not instrumented by accident.
- [ ] This is something else
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)

### CODE REVIEW

File: apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java [hunk 1]
```diff
@@ -182,9 +190,47 @@ public Span startSpan() {
...
+        byte[] header = new byte[TraceContext.BINARY_FORMAT_EXPECTED_LENGTH];
+        for (int i = 0; i < links.size(); i+=2) {
+            byte[] traceID = links.get(i);
+            byte[] spanID = links.get(i+1);
+            TraceContext.filIdOutgoingTraceParentBinaryHeader(header, traceID, spanID);
+            span.addSpanLink(
+                TraceContext.getFromTraceContextBinaryHeaders(),
+                BinaryByteArrayAccessor.INSTANCE,
+                header);
+        }
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java:206]: I think this is forcing complicated logic (and allocation) only to match the current `AbstractSpan#addSpanLink` API, which assumes the caller is dealing with a header carrier and a dedicated accessor that knows how to get it based on name.
However, if you add other `AbstractSpan#addSpanLink()` APIs, that could get much simpler.

When we looked together, I thought you could add `AbstractSpan#addSpanLink(String/byte[] traceId, String/byte[] spanId)`, that could be, in which case the implementation would grab a `TraceContext` from the object pool and fill it with the required details (probably requiring a new `TraceContext#fromParent(String/byte[] traceId, String/byte[] parentId)`).

However, now that I looked at it again, it seems it may get even simpler than that - the `OTelSpanBuilder#addLink()` APIs receive a `io.opentelemetry.api.trace.SpanContext` as the parent. IIANM, this is always expected to be our implementation - `co.elastic.apm.agent.opentelemetry.sdk.OTelSpanContext`, which holds a `TraceContext` already.  So if you add an `AbstractSpan#addSpanLink(TraceContext parent)`, you're basically done. Since `co.elastic.apm.agent.opentelemetry.sdk.OTelSpanContext` seems immutable and non recyclable, you can add the `TraceContext` it refers to directly to the links list. Otherwise, if you fear it could be leaked, you can take a new `TraceContext` from the object pool and fill it through `copyFrom`.

On the other hand, I didn't really try anything, so orry if I am missing something....

File: apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java [hunk 2]
```diff
@@ -182,9 +190,47 @@ public Span startSpan() {
...
+        links.clear();
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java:207]: Why is this required?

**Review by eyalkoren — COMMENTED** [review]
See my comment inside- I think it can get more straightforward.
If I am wrong, I apologize.

File: apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java (hunk 1 above)
Comment (eyalkoren) [reply apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java:206]: I just realized that since the span links are saved as children of the links, you cannot use the context itself. 
So the better API to add to `AbstractSpan` would be: 
```
public <T> boolean addSpanLink(ChildContextCreator<T> creator, T parent)
```
Of which implementation would still be extremely simple and very similar to the existing `addSpanLink`:
1. grab a `TraceContext` instance from the pool
2. initialize it through `creator.asChildOf()`

And to use it with a `TraceContext` you would do:
```
span.addSpanLink(TraceContext.fromParentContext(), parentTraceContext)
```

However, criticizing my own work- keeping the span links as children of the linked spans was a compromise of doing something less intuitive in order to avoid complicating `TraceContext`. However, I now think it makes sense to add the ability to create `TraceContext` from headers not as children and makes things more straightforward.
Let's talk about it...

Comment (jackshirazi) [reply apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java]: I've applied that advice (thanks!), have a new look

File: apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java [hunk 3]
```diff
@@ -182,6 +188,11 @@ public Span startSpan() {
         // user outcome to provide higher priority and avoid inferring outcome from any reported exception
         span.withUserOutcome(Outcome.UNKNOWN);
 
+        // Add the links to the span
+        for (int i = 0; i < links.size(); i++) {
+            span.addSpanLink(TraceContext.fromParentContext(), links.get(i));
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java:193]: We can delay the use of `((OTelSpanContext) spanContext).getElasticTraceContext()` until here.
I think it better reflects who's responsible to what - the user is responsible to make sure that at this point all OTel states are correct and it is our responsibility to take it from here.

File: apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java [hunk 4]
```diff
@@ -86,13 +91,14 @@ public SpanBuilder setNoParent() {
...
+        links.add(((OTelSpanContext) spanContext).getElasticTraceContext());
         return this;
     }
 
     @Override
-    public SpanBuilder addLink(SpanContext spanContext, Attributes attributes) {
-        addLinkLogger.warn("The addLink API is not supported at the moment");
+    public SpanBuilder addLink(SpanContext spanContext, Attributes attributes1) {
+        addLink(spanContext);
+        attributes1.forEach((AttributeKey<?> k, Object v) -> attributes.put((AttributeKey<? super Object>) k, (Object) v));
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java:101]: Is that the intention? In the OTel javadoc it only says: "the attributes of the Link". I don't know what this means - are those intended to be added to the built span attributes?

**Review by eyalkoren — COMMENTED** [review]
Thanks for applying the suggestion, looks good!
Only made a tiny proposal and a question about link attributes

File: apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java (hunk 4 above)
Comment (felixbarny) [reply apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java:101]: These shouldn't be added to the span attributes as they're scoped to a particular span link. However, we don't have support for span link attributes (an no OTel auto instrumentation seems to use span link attributes yet). So we can just ignore the span link attributes for now.

Comment (jackshirazi) [reply apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java:101]: In this case I think this method should revert to the API not supported rather than add the link and ignore the attributes. That way it's clear to the user they shouldn't use it

File: apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java [hunk 5]
```diff
@@ -58,6 +61,8 @@ class OTelSpanBuilder implements SpanBuilder {
     @Nullable
     private Context remoteContext;
 
+    private List<TraceContext> links = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (jackshirazi) [reply apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java]: okay, will do

File: apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java (hunk 3 above)
Comment (jackshirazi) [reply apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java]: agreed

File: apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java (hunk 4 above)
Comment (eyalkoren) [reply apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java:101]: The only down side to that is that users the use it currently with `null` or empty attributes won't be able to migrate to using us.
What about logging a warning once, saying these attributes will be ignored?

Comment (jackshirazi) [reply apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-plugin/src/main/java/co/elastic/apm/agent/opentelemetry/sdk/OTelSpanBuilder.java]: Yes, agree. I went with logging the warning every time, it's what we did previously and better to make that clear alongside the call when investigating

**Review by eyalkoren — APPROVED** [review]
